components/builder-depot/src/server.rs

```diff
@@ -461,32 +466,76 @@ fn download_latest_origin_encryption_key(req: &mut Request) -> IronResult<Respon
 }
 
-fn generate_origin_encryption_keys(req: &mut Request) -> IronResult<Response> {
-    debug!("Generate Origin Encryption Keys {:?}", req);
-    match get_param(req, "origin") {
-        Some(origin) => {
-            if !check_origin_access(req, &origin).unwrap_or(false) {
-                return Ok(Response::with(status::Forbidden));
-            }
+fn generate_origin_encryption_keys(
+    origin: &str,
+    req: &mut Request,
+) -> Result<OriginPublicEncryptionKey> {
+    debug!("Generate Origin Encryption Keys {:?} for {:?}", req, origin);
 
-            match helpers::get_origin(req, origin) {
-                Ok(origin) => {
-                    match helpers::generate_origin_encryption_keys(req, origin) {
-                        Ok(_) => Ok(Response::with(status::Created)),
-                        Err(err) => Ok(render_net_error(&err)),
-                    }
-                }
-                Err(err) => Ok(render_net_error(&err)),
+    match helpers::get_origin(req, origin) {
+        Ok(origin) => {
+            let mut public_request = OriginPublicEncryptionKeyCreate::new();
+            let mut private_request = OriginPrivateEncryptionKeyCreate::new();
+            let mut public_key = OriginPublicEncryptionKey::new();
+            let mut private_key = OriginPrivateEncryptionKey::new();
+            {
+                let session = req.extensions.get::<Authenticated>().unwrap();
+                public_key.set_owner_id(session.get_id());
+                private_key.set_owner_id(session.get_id());
             }
+            public_key.set_name(origin.get_name().to_string());
+            public_key.set_origin_id(origin.get_id());
+            private_key.set_name(origin.get_name().to_string());
+            private_key.set_origin_id(origin.get_id());
+
+            let pair = BoxKeyPair::generate_pair_for_origin(origin.get_name())
+                .expect("failed to generate origin encryption key pair");
+            public_key.set_revision(pair.rev.clone());
+            public_key.set_body(
+                pair.to_public_string()
+                    .expect("no public key in generated encryption key pair")
+                    .into_bytes(),
+            );
+            private_key.set_revision(pair.rev.clone());
+            private_key.set_body(
+                pair.to_secret_string()
+                    .expect("no secret key in generated encryption key pair")
+                    .into_bytes(),
+            );
+
+            public_request.set_public_encryption_key(public_key);
+            private_request.set_private_encryption_key(private_key);
+
+            let key =
+                route_message::<OriginPublicEncryptionKeyCreate, OriginPublicEncryptionKey>(
+                    req,
+                    &public_request,
+                )?;
+            route_message::<OriginPrivateEncryptionKeyCreate, OriginPrivateEncryptionKey>(
+                req,
+                &private_request,
+            )?;
```

Comment:
What happens if the `OriginPublicEncryptionKeyCreate` succeeds but the `OriginPrivateEncryptionKeyCreate` fails?  Is the user left in a bad state?  If they try to create a new key again, will it succeed if this was aborted in a partially finished state?